File: src/utils/webring.ts
interface WebringURL {
  name: string;
  url: string;
}

interface WebringURLs {
  prev: WebringURL;
  next: WebringURL;
}

export default async (): Promise<WebringURLs> => {
  const res = await fetch(
    "https://pombosmalvados.github.io/webring/webring.json",
  );
  const websites: WebringURL[] = await res.json();

  const index = websites.findIndex(
    (website) => website.name === "Eduardo Espadeiro",
  );
  const prevIndex = (index + websites.length - 1) % websites.length;
  const nextIndex = (index + 1) % websites.length;
  const prev = websites[prevIndex];
  const next = websites[nextIndex];

  return {
    prev,
    next,
  };
};

File: src/webring.rs
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct Website {
    name: String,
    url: String,
}

fn fetch_webring_members() -> Result<Vec<Website>, ureq::Error> {
    let mut response = ureq::get("https://pombosmalvados.github.io/webring/webring.json").call()?;
    let body = response.body_mut().read_json()?;
    Ok(body)
}

pub fn fetch_neighbors() -> Option<(Website, Website)> {
    let webring = fetch_webring_members().ok()?;
    let pos = webring
        .iter()
        .position(|website| website.name == "Eduardo Espadeiro")?;
    let prev = webring[(webring.len() + pos - 1) % webring.len()].clone();
    let next = webring[(pos + 1) % webring.len()].clone();
    Some((prev, next))
}
